include/executors/blas_packet_traits_sycl.hpp

```diff
@@ -0,0 +1,124 @@
+/***************************************************************************
+ *
+ *  @license
+ *  Copyright (C) 2016 Codeplay Software Limited
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL-BLAS: BLAS implementation using SYCL
+ *
+ *  @filename blas_packet_traits_sycl.hpp
+ *
+ **************************************************************************/
+
+#ifndef BLAS_PACKET_TRAITS_SYCL_HPP
+#define BLAS_PACKET_TRAITS_SYCL_HPP
+
+#include <CL/sycl.hpp>
+
+#include <CL/sycl.hpp>
+#include <executors/blas_packet_traits.hpp>
+
+namespace blas {
+
+class SYCLDevice {
+  /*!
+   * @brief SYCL queue for execution of trees.
+   */
+  cl::sycl::device m_device;
+  /*!
+   * @brief SYCL device.
+   */
+  cl::sycl::queue m_queue;
+public:
+  SYCLDevice():
+    m_queue(cl::sycl::queue([&](cl::sycl::exception_list l) {
+      for(const auto& e : l) {
+        try {
+          if(e) {
+            std::rethrow_exception(e);
+          }
+        } catch(cl::sycl::exception e) {
+          std::cerr << e.what() << std::endl;
+        }
+      }
+    })),
+    m_device(m_queue.get_device())
+  {}
+
+  static void parallel_for_setup(size_t &localSize, size_t &nWG, size_t globalSize = UINT_MAX) {
+    localSize = 256;
+    nWG = 512;
+    if(globalSize != UINT_MAX) {
+      nWG = (nWG + (2 * localSize) - 1) / (2 * localSize);
+    }
+  }
+
+  cl::sycl::queue &sycl_queue() {
```

Comment:
Returning a copy of the object is safer in this case, since its reference counted. This means the only thing you copy is a thin object that has only a pointer to the actual implementation. In ComputeCpp is implemented as http://en.cppreference.com/w/cpp/language/pimpl